File: baekjoon/BOJ_2661.java
package backtracking;

import java.util.Scanner;

public class BOJ_2661 {

    static int n;

    static int answer = Integer.MAX_VALUE;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        n = scanner.nextInt();
        if (n == 1) {
            System.out.println(1);
        }
        backtracking("");

    }

    static void backtracking(String s) {
        if (n == s.length()) {
            System.out.println(s);
            System.exit(0);
        }
        for (int i = 1; i <= 3; i++) {
            if (isGood(s + i))
                backtracking(s + i);
        }
    }

    static boolean isGood(String s) {
        int len = s.length() / 2;
        for (int i = 1; i <= len; i++) {
            if (s.substring(s.length() - i).equals(s.substring(s.length() - 2 * i, s.length() - i))) {
                return false;
            }
        }
        return true;
    }
}

File: baekjoon/BOJ_2661.kt
package boj.backtracking

import kotlin.system.exitProcess

private fun main() {
    val n = readLine()!!.toInt()
    if (n == 1) {
        println(1)
        return
    }
    backtracking("", n)

}

private fun backtracking(s: String, n: Int) {
    if (s.length == n) {
        println(s)
        exitProcess(0)
    }
    for (i in 1..3) {
        if (isGood(s + i)) {
            backtracking(s + i, n)
        }
    }
}

private fun isGood(s: String): Boolean {
    var len = s.length / 2
    for (i in 1..len) {
        if (s.substring(s.length - i) == s.substring(s.length - 2 * i, s.length - i)) {
            return false;
        }
    }
    return true
}